[lief] Add new port for lief-project/LIEF

BehroozAbbassi opened the pull request:
**Describe the pull request**
Adds the [lief](https://github.com/lief-project/LIEF) library into the vcpkg ports

- #### What does your PR fix?  
  Fixes #...

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

BehroozAbbassi committed:
[lief]  Add new port for lief-project/LIEF

BehroozAbbassi committed:
[lief]  Add new port for lief-project/LIEF

JonLiu1993 commented:
@BehroozAbbassi ,Thanks for your pr, could you please take a look
https://dev.azure.com/vcpkg/public/_build/results?buildId=62535&view=artifacts&pathAsName=false&type=publishedArtifacts

JonLiu1993 commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,97 @@
+        -DLIEF_USE_CRT_RELEASE=MD
+)
+
+vcpkg_install_cmake()
```
```suggestion
vcpkg_cmake_install()
```

JonLiu1993 commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,97 @@
+vcpkg_copy_pdbs()
+
+# # Handle copyright
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/LIEF RENAME copyright)
```
```suggestion
file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/LIEF" RENAME copyright)
```

JonLiu1993 commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,97 @@
+# # Handle copyright
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/LIEF RENAME copyright)
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
```
```suggestion
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
```

JonLiu1993 commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,97 @@
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/LIEF RENAME copyright)
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/share)
```
```suggestion
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
```

JonLiu1993 commented on ports/lief/vcpkg.json:
```diff
@@ -0,0 +1,96 @@
+{
+  "name": "lief",
+  "version-string": "0.11.5",
```
```suggestion
  "version": "0.11.5",
```

JonLiu1993 commented on ports/lief/vcpkg.json:
```diff
@@ -0,0 +1,96 @@
+{
+  "name": "lief",
+  "version-string": "0.11.5",
+  "description": "LIEF - Library to Instrument Executable Formats",
+  "homepage": "https://lief.quarkslab.com",
+  "default-features": [
```
```suggestion
  "description": "LIEF - Library to Instrument Executable Formats",
  "homepage": "https://lief.quarkslab.com",
  "dependencies": [
    {
      "name": "vcpkg-cmake",
      "host": true
    },
    {
      "name": "vcpkg-cmake-config",
      "host": true
    }
  ],
  "default-features": [
```

JonLiu1993 commented:
@BehroozAbbassi ,Thanks for your pr,could you please take a look:
https://dev.azure.com/vcpkg/public/_build/results?buildId=62535&view=artifacts&pathAsName=false&type=publishedArtifacts

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Update ports/lief/vcpkg.json

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Update ports/lief/vcpkg.json

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Update portfile.cmake

BehroozAbbassi committed:
Update CI baseline

BehroozAbbassi committed:
Update portfile.cmake

BehroozAbbassi committed:
Update portfile.cmake

BehroozAbbassi commented:
@JonLiu1993 Sorry for the delay :(

I'm new to VCPKG contribution and just can't figure out how to fix this problem
how exactly should I add en expected error to the `.\scripts\ci.baseline.txt` ? manually or it has a command?

```
REGRESSION: lief:x64-linux. If expected, add lief:x64-linux=fail to .\scripts\ci.baseline.txt.
```

JonLiu1993 commented:
> @JonLiu1993 Sorry for the delay :(
> 
> I'm new to VCPKG contribution and just can't figure out how to fix this problem how exactly should I add en expected error to the `.\scripts\ci.baseline.txt` ? manually or it has a command?
> 
> ```
> REGRESSION: lief:x64-linux. If expected, add lief:x64-linux=fail to .\scripts\ci.baseline.txt.
> ```

This is not a modification suggestion for you, please run the command "./vcpkg x-add-version lief --overwrite-version " and commit again

JonLiu1993 committed:
update version

JonLiu1993 commented:
@BehroozAbbassi ,This is detail error log:
https://dev.azure.com/vcpkg/public/_build/results?buildId=63521&view=artifacts&pathAsName=false&type=publishedArtifacts

BehroozAbbassi commented:
> 

I really confused, I used the `vcpkg_fail_port_install` to indicate that library can't be build with the `static-md` triplet, **so definitely there will be an error** with this triplet!
how should I pass the CI with this error?


```cmake
if ("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static" )
    if ("${VCPKG_CRT_LINKAGE}" STREQUAL "dynamic" )
        vcpkg_fail_port_install(MESSAGE "${PORT} doesn't support static-md" ALWAYS)
    endif()
endif()
```

error message in `failure logs for x64-windows-static-md`

```
CMake Error at scripts/cmake/vcpkg_fail_port_install.cmake:96 (message):
  lief doesn't support static-md

Call Stack (most recent call first):
  ports/lief/portfile.cmake:5 (vcpkg_fail_port_install)
  scripts/ports.cmake:142 (include)

```

PhoebeHui commented:
@BehroozAbbassi, You can add lief:x64-windows-static-md=fail  to https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt.

JonLiu1993 commented:
@BehroozAbbassi ,Have you tried phoebe's review suggestions?

JonLiu1993 commented:
@BehroozAbbassi ,Ping for response

BehroozAbbassi commented:
> @BehroozAbbassi ,Have you tried phoebe's review suggestions?

That's my question, https://github.com/microsoft/vcpkg/pull/21213#issuecomment-974704165
should I add it to the text file manually or it has a command or something for it?

BehroozAbbassi committed:
Merge branch 'microsoft:master' into master

BehroozAbbassi committed:
Add MSVC specific variables to the LIEF_USE_CRT_RELEASE

BehroozAbbassi committed:
add lief:x64-windows-static-md=fail to the ci.baseline.txt

BehroozAbbassi committed:
Update version database

BehroozAbbassi committed:
update ci.baseline.txt

BehroozAbbassi commented:
> 

So I manually add these exceptions to the `ci.baseline.txt` file to ignore that errors, but I'm not sure about the **None Windows** errors like `lief:x64-linux=fai` because I only work on windows

```
# LIEF library
lief:x64-windows-static-md=fail
lief:x64-linux=fail
lief:x64-uwp=fail
lief:arm-uwp=fail
lief:x64-osx=fail
```

JonLiu1993 commented:
> > @BehroozAbbassi ,Have you tried phoebe's review suggestions?
> 
> That's my question, [#21213 (comment)](https://github.com/microsoft/vcpkg/pull/21213#issuecomment-974704165) should I add it to the text file manually or it has a command or something for it?

@BehroozAbbassi,We need to manually add the triplet it fails in ci.baseline.txt

JonLiu1993 commented on scripts/ci.baseline.txt:
```diff
@@ -533,6 +533,14 @@ libcopp:arm-uwp=fail
 # Missing system libraries on linux to run/prepare autoconf
 libgpod:x64-linux=fail
 libgpod:x64-osx=fail
+
```
```suggestion
```

JonLiu1993 commented on scripts/ci.baseline.txt:
```diff
@@ -533,6 +533,14 @@ libcopp:arm-uwp=fail
+lief:x64-uwp=fail
+lief:arm-uwp=fail
+lief:x64-osx=fail
+
```
```suggestion
```

JonLiu1993 commented on scripts/ci.baseline.txt:
```diff
@@ -533,6 +533,14 @@ libcopp:arm-uwp=fail
+
+# LIEF library
+lief:x64-windows-static-md=fail
+lief:x64-linux=fail
+lief:x64-uwp=fail
+lief:arm-uwp=fail
+lief:x64-osx=fail
```
And if this port doesn't support some triplets, please add supports keyword in vcpkg.json instead of modify baseline.
Only ports that do not support certain triples in vcpkg for the time being can write them to the baseline.

BehroozAbbassi committed:
Update scripts/ci.baseline.txt

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Update scripts/ci.baseline.txt

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Add supports platforms/triplets to the vcpkg.json

BehroozAbbassi committed:
Update version database

BehroozAbbassi committed:
Update ci.baseline.txt

BehroozAbbassi committed:
Update version database

JonLiu1993 commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,121 @@
+endif()
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA # Disable this option if project cannot be built with Ninja
```
```suggestion
    SOURCE_PATH "${SOURCE_PATH}"
```

JonLiu1993 commented on ports/lief/vcpkg.json:
```diff
@@ -0,0 +1,107 @@
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
```
```suggestion
```

JonLiu1993 replied:
This dependency is not actually needed

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Update ports/lief/vcpkg.json

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

BehroozAbbassi committed:
Update version database

JackBoosY commented:
> > 
> 
> I really confused, I used the `vcpkg_fail_port_install` to indicate that library can't be build with the `static-md` triplet, **so definitely there will be an error** with this triplet! how should I pass the CI with this error?
> 
> ```cmake
> if ("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static" )
>     if ("${VCPKG_CRT_LINKAGE}" STREQUAL "dynamic" )
>         vcpkg_fail_port_install(MESSAGE "${PORT} doesn't support static-md" ALWAYS)
>     endif()
> endif()
> ```
> 
> error message in `failure logs for x64-windows-static-md`
> 
> ```
> CMake Error at scripts/cmake/vcpkg_fail_port_install.cmake:96 (message):
>   lief doesn't support static-md
> 
> Call Stack (most recent call first):
>   ports/lief/portfile.cmake:5 (vcpkg_fail_port_install)
>   scripts/ports.cmake:142 (include)
> ```

But why? Can you please provide the failure log?  Also please provide the Linux failure logs.

Thanks.

JonLiu1993 commented:
@BehroozAbbassi ,Ping for response

ekilmer commented:
I am happy to see this project being packaged for vcpkg.

I was going to attempt this myself before I searched for an existing PR/issue. I know LIEF requires some external dependencies, and they've listed them here https://lief-project.github.io/doc/latest/compilation.html#third-party but I don't see any of these dependencies listed in the `vcpkg.json` file.

ekilmer commented:
Looks like the following CMake variables can enable `find_package` on the external dependencies:
* `-DLIEF_OPT_MBEDTLS_EXTERNAL=ON` for `mbedtls` (required)
* `-DLIEF_OPT_NLOHMANN_JSON_EXTERNAL=ON` for `nlohmann-json` (feature: `enable-json`)
* `-DLIEF_OPT_UTFCPP_EXTERNAL=ON` for `utfcpp` (required)
* `-DLIEF_EXTERNAL_SPDLOG=ON` for `spdlog` (required)

and some others that I need to look closer at (will update this comment if I figure it out)
* `-DLIEF_EXTERNAL_LEAF=ON` for Boost Leaf
* Can't find an option to use external find_package for `frozen`

romainthomas commented:
@ekilmer Yes the *external* configuration is not released yet. And yes, I missed to provide the option for Frozen. 
I'll add it. Let me know if it needs other *tweaks* in the build system before releasing ``LIEF v0.12.0``.

BehroozAbbassi commented:
> > > 
> > 
> > 
> > I really confused, I used the `vcpkg_fail_port_install` to indicate that library can't be build with the `static-md` triplet, **so definitely there will be an error** with this triplet! how should I pass the CI with this error?
> > ```cmake
> > if ("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static" )
> >     if ("${VCPKG_CRT_LINKAGE}" STREQUAL "dynamic" )
> >         vcpkg_fail_port_install(MESSAGE "${PORT} doesn't support static-md" ALWAYS)
> >     endif()
> > endif()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > error message in `failure logs for x64-windows-static-md`
> > ```
> > CMake Error at scripts/cmake/vcpkg_fail_port_install.cmake:96 (message):
> >   lief doesn't support static-md
> > 
> > Call Stack (most recent call first):
> >   ports/lief/portfile.cmake:5 (vcpkg_fail_port_install)
> >   scripts/ports.cmake:142 (include)
> > ```
> 
> But why? Can you please provide the failure log? Also please provide the Linux failure logs.
> 
> Thanks.

BehroozAbbassi commented:
> > > 
> > 
> > 
> > I really confused, I used the `vcpkg_fail_port_install` to indicate that library can't be build with the `static-md` triplet, **so definitely there will be an error** with this triplet! how should I pass the CI with this error?
> > ```cmake
> > if ("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static" )
> >     if ("${VCPKG_CRT_LINKAGE}" STREQUAL "dynamic" )
> >         vcpkg_fail_port_install(MESSAGE "${PORT} doesn't support static-md" ALWAYS)
> >     endif()
> > endif()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > error message in `failure logs for x64-windows-static-md`
> > ```
> > CMake Error at scripts/cmake/vcpkg_fail_port_install.cmake:96 (message):
> >   lief doesn't support static-md
> > 
> > Call Stack (most recent call first):
> >   ports/lief/portfile.cmake:5 (vcpkg_fail_port_install)
> >   scripts/ports.cmake:142 (include)
> > ```
> 
> But why? Can you please provide the failure log? Also please provide the Linux failure logs.
> 
> Thanks.

Sorry for the delay, I guess the problem is solved now. I don't know how to get the failure logs of that specific commit :( I think I was missing something at that point, by now everything is OK.

JackBoosY commented:
Can you please resolve the file conflict?

Thanks.

BehroozAbbassi committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

BehroozAbbassi committed:
Merge branch 'master' into master

BehroozAbbassi committed:
Merge branch 'master' of https://github.com/BehroozAbbassi/vcpkg

ekilmer commented:
@romainthomas I've opened a draft PR (#22957) for testing/experimenting with packaging the latest commit (at time of writing) on `master` branch of LIEF.

There are two patches that I thought were necessary:

1. CMake fixes for correct packaging/distribution (`ports/lief/support_distribution.patch`)
2. Fixing a Windows compilation issue (`ports/lief/fix_windows_build.patch`) using the workaround from this issue https://github.com/lief-project/LIEF/issues/616.

It would probably make sense to further discuss the patch contents on the #22957 PR

JackBoosY commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,120 @@
+
+if ("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static" )
+    if ("${VCPKG_CRT_LINKAGE}" STREQUAL "dynamic" )
+        vcpkg_fail_port_install(MESSAGE "${PORT} doesn't support static-md" ALWAYS)
+    endif()
+endif()
+
```
```suggestion
```

JackBoosY replied:
Why this port doesn't support static-md?

BehroozAbbassi replied:
> 

I don't remember the exact reason, but it's related to the MSVC's CRT 
For static build we must use '/MT' so there is no place for '/MD' in the static build!

BehroozAbbassi replied:
> Why this port doesn't support static-md?

I don't remember the exact reason, but it's related to the MSVC's CRT
For static build we must use '/MT' so there is no place for '/MD' in the static build!

JackBoosY replied:
Can you please patch the configure files to remove the CRT set?

BehroozAbbassi replied:
> Can you please patch the configure files to remove the CRT set?

Could you explain why we should remove it? It has problems with the wrong `Runtime Library` settings that's why I did it like this.

JackBoosY replied:
`vcpkg_fail_port_install` was already deprecated, we should use field `supports` in `vcpkg.json`.
And we should know why static-md is not supported.

BehroozAbbassi replied:
> 

I already add the `supports` field in `vcpkg.json` but it's not descriptive enough, it's just a cryptic expression `((x86 | x64) & ((windows & static) | windows)) & !uwp` :)

But I guess it's much better than using a deprecated functionality!
I'll test it tomorrow 


But about **the reason**, as I said by using static build we use the `/MT` otherwise (dynamic build) we use the `/MD', so these two can't combine to gather.

Actually I didn't test it recently but i has the problem with CRT for long time @romainthomas could we have such a build?

JackBoosY replied:
If the upstream doesn't support static-md, we should disable it in `supports`, not here.

BehroozAbbassi replied:
> If the upstream doesn't support static-md, we should disable it in `supports`, not here.

I couldn't figure out how to disable `static-md` in the `supports`!
In the `supports` expressions we can only use `static` or `!static` that is an indicator for the `VCPKG_LIBRARY_LINKAGE ` and not `VCPKG_CRT_LINKAGE`.

To manage the `md` we need to control `VCPKG_CRT_LINKAGE` too! any idea?

JackBoosY replied:
Such as: `"!(windows & static & !staticcrt)"`

JackBoosY commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,120 @@
+    "tests"          LIEF_TESTS             # Enable tests
+    "doc"            LIEF_DOC               # Enable documentation
+    "python-api"     LIEF_PYTHON_API        # Enable Python API
+    "install-python" LIEF_INSTALL_PYTHON    # Install Python bindings
+    "c-api"          LIEF_C_API             # C API
+    "examples"       LIEF_EXAMPLES          # Build LIEF C++ examples
+    "force32"        LIEF_FORCE32           # Force build LIEF 32 bits version
+    "coverage"       LIEF_COVERAGE          # Perform code coverage
+    "use-ccache"     LIEF_USE_CCACHE        # Use ccache to speed up compilation
+    "extra-warnings" LIEF_EXTRA_WARNINGS    # Enable extra warning from the compiler    
+    "logging"        LIEF_LOGGING           # Enable logging
+    "logging-debug"  LIEF_LOGGING_DEBUG     # Enable debug logging
+    "enable-json"    LIEF_ENABLE_JSON       # Enable JSON-related APIs
+    "shared-lib"     LIEF_SHARED_LIB        # Enable shared lib
+
+    # "disable_frozen" LIEF_DISABLE_FROZEN    # Disable Frozen even if it is supported
+
+    "elf"            LIEF_ELF               # Build LIEF with ELF module
+    "pe"             LIEF_PE                # Build LIEF with PE  module
+    "macho"          LIEF_MACHO             # Build LIEF with MachO module
+    
+    "oat"            LIEF_OAT               # Build LIEF with OAT module
+    "dex"            LIEF_DEX               # Build LIEF with DEX module
+    "vdex"           LIEF_VDEX              # Build LIEF with VDEX module
+    "art"            LIEF_ART               # Build LIEF with ART module
+
+    # Sanitizer
+    "asan"          LIEF_ASAN               # Enable Address sanitizer
+    "lsan"          LIEF_LSAN               # Enable Leak sanitizer
+    "tsan"          LIEF_TSAN               # Enable Thread sanitizer
+    "usan"          LIEF_USAN               # Enable undefined sanitizer
+
+    # Fuzzer
+    "fuzzing"       LIEF_FUZZING            # Fuzz LIEF
+        
+    # Profiling
+    "profiling"     LIEF_PROFILING          # Enable performance profiling
+
+INVERTED_FEATURES
+    "disable-frozen" LIEF_DISABLE_FROZEN    # Disable Frozen even if it is supported
```
```suggestion
    FEATURES 
        "tests"          LIEF_TESTS             # Enable tests
        "doc"            LIEF_DOC               # Enable documentation
        "python-api"     LIEF_PYTHON_API        # Enable Python API
        "install-python" LIEF_INSTALL_PYTHON    # Install Python bindings
        "c-api"          LIEF_C_API             # C API
        "examples"       LIEF_EXAMPLES          # Build LIEF C++ examples
        "force32"        LIEF_FORCE32           # Force build LIEF 32 bits version
        "coverage"       LIEF_COVERAGE          # Perform code coverage
        "use-ccache"     LIEF_USE_CCACHE        # Use ccache to speed up compilation
        "extra-warnings" LIEF_EXTRA_WARNINGS    # Enable extra warning from the compiler    
        "logging"        LIEF_LOGGING           # Enable logging
        "logging-debug"  LIEF_LOGGING_DEBUG     # Enable debug logging
        "enable-json"    LIEF_ENABLE_JSON       # Enable JSON-related APIs
        "shared-lib"     LIEF_SHARED_LIB        # Enable shared lib
    
        # "disable_frozen" LIEF_DISABLE_FROZEN    # Disable Frozen even if it is supported
    
        "elf"            LIEF_ELF               # Build LIEF with ELF module
        "pe"             LIEF_PE                # Build LIEF with PE  module
        "macho"          LIEF_MACHO             # Build LIEF with MachO module
    
        "oat"            LIEF_OAT               # Build LIEF with OAT module
        "dex"            LIEF_DEX               # Build LIEF with DEX module
        "vdex"           LIEF_VDEX              # Build LIEF with VDEX module
        "art"            LIEF_ART               # Build LIEF with ART module
    
        # Sanitizer
        "asan"          LIEF_ASAN               # Enable Address sanitizer
        "lsan"          LIEF_LSAN               # Enable Leak sanitizer
        "tsan"          LIEF_TSAN               # Enable Thread sanitizer
        "usan"          LIEF_USAN               # Enable undefined sanitizer
    
        # Fuzzer
        "fuzzing"       LIEF_FUZZING            # Fuzz LIEF
    
        # Profiling
        "profiling"     LIEF_PROFILING          # Enable performance profiling
    
    INVERTED_FEATURES
        "disable-frozen" LIEF_DISABLE_FROZEN    # Disable Frozen even if it is supported
```
Just add 4 space for format.

JackBoosY commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,120 @@
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+
```
```suggestion
```

JackBoosY commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,120 @@
+        ${FEATURE_OPTIONS}
+    OPTIONS_DEBUG    
+        -DLIEF_USE_CRT_DEBUG=${LIEF_CRT_DEBUG}
+
```
```suggestion
```

JackBoosY commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,120 @@
+
+    OPTIONS_RELEASE
+        -DLIEF_USE_CRT_RELEASE=${LIEF_CRT_RELEASE}
+    
```
```suggestion
```

JackBoosY commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,120 @@
+
+vcpkg_cmake_install()
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
+    if(VCPKG_TARGET_IS_WINDOWS)
+        if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
+            if(EXISTS "${CURRENT_PACKAGES_DIR}/debug/lib/LIEF.dll")
+                file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/debug/bin")
+                file(RENAME "${CURRENT_PACKAGES_DIR}/debug/lib/LIEF.dll" "${CURRENT_PACKAGES_DIR}/debug/bin/LIEF.dll")
+            endif()
+        endif()
+        if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
+            if(EXISTS "${CURRENT_PACKAGES_DIR}/lib/LIEF.dll")
+                file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/bin")
+                file(RENAME "${CURRENT_PACKAGES_DIR}/lib/LIEF.dll" "${CURRENT_PACKAGES_DIR}/bin/LIEF.dll")
+            endif()
+        endif()
+    endif()
+endif()
```
I think we should make a patch to implement this.

BehroozAbbassi replied:
> I think we should make a patch to implement this.

you mean patch the source tree?
I'm so new to the vcpk's cmake, so i'm not sure what do you mean exactly.

JackBoosY replied:
Yeah, patch the source.

BehroozAbbassi replied:
> Yeah, patch the source.

I've no idea about this one, does this `cmake` code have any problem?

JackBoosY replied:
The upstream cmake code doesn't have any problems, but it doesn't suitable for vcpkg.
In vcpkg, all the dynamic libraries and pdbs need to be installed to _bin_, so what you need to do is only modify the cmake install code, example:
```diff
install(TARGETS xxx
-RUNTIME DESTINATION lib
+RUNTIME DESTINATION bin
...
)
```

JackBoosY replied:
Ping.

BehroozAbbassi replied:
> The upstream cmake code doesn't have any problems, but it doesn't suitable for vcpkg. In vcpkg, all the dynamic libraries and pdbs need to be installed to _bin_, so what you need to do is only modify the cmake install code, example:
> 
> ```diff
> install(TARGETS xxx
> -RUNTIME DESTINATION lib
> +RUNTIME DESTINATION bin
> ...
> )
> ```

Sorry I missed this one, I just did it.
https://github.com/microsoft/vcpkg/pull/21213/commits/36b749fe63b9c770282c9ec7a47951f17cf08e92

JackBoosY commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,120 @@
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
+
+message(STATUS "[LIEF DEBUG] Installing done")
```
```suggestion
```

JackBoosY commented on scripts/ci.baseline.txt:
```diff
@@ -482,6 +482,11 @@ libcopp:arm-uwp=fail
 # Missing system libraries on linux to run/prepare autoconf
 libgpod:x64-linux=fail
 libgpod:x64-osx=fail
+# LIEF library
+lief:x64-windows-static-md=fail
```
```suggestion
```

JackBoosY commented on scripts/ci.baseline.txt:
```diff
@@ -482,6 +482,11 @@ libcopp:arm-uwp=fail
 libgpod:x64-osx=fail
+# LIEF library
+lief:x64-windows-static-md=fail
+cpuid:arm-uwp=fail
+cpuid:x64-uwp=fail
+cpuid:arm64-windows=fail
```
Why add this?

BehroozAbbassi replied:
It was the conflict! some one else add these to the `ci.baseline.txt`

BehroozAbbassi replied:
> Why add this?

It was the conflict! some one else add these to the `ci.baseline.txt`

JackBoosY replied:
This record was removed in PR https://github.com/microsoft/vcpkg/pull/22735, please also remove them.

BehroozAbbassi replied:
> This record was removed in PR #22735, please also remove them.

Sorry my bad, removed it now.

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

BehroozAbbassi committed:
Merge branch 'master' of https://github.com/BehroozAbbassi/vcpkg

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

BehroozAbbassi committed:
Merge branch 'master' of https://github.com/BehroozAbbassi/vcpkg

BehroozAbbassi committed:
Update version database

BehroozAbbassi committed:
Update ci.baseline.txt

JonLiu1993 commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,115 @@
+
+if ("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static" )
+    if ("${VCPKG_CRT_LINKAGE}" STREQUAL "dynamic" )
+        vcpkg_fail_port_install(MESSAGE "${PORT} doesn't support static-md" ALWAYS)
+    endif()
+endif()
+
```
```suggestion
```

JonLiu1993 replied:
@BehroozAbbassi , If the upstream doesn't support static-md, we should disable it in supports, not here. Such as: "!(windows & static & !staticcrt)"

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

BehroozAbbassi committed:
Update ci and change supports to ignore "static-md"

BehroozAbbassi committed:
Update version database

BehroozAbbassi committed:
Update version database

BehroozAbbassi committed:
Update portfile.cmake

BehroozAbbassi committed:
Update version database

JonLiu1993 commented:
@BehroozAbbassi ,When I tested the features on my local machine with command "./vcpkg install lief[*] " and "./vcpkg install lief[*]:x64-windows" I met the error log:
````
CMake Error at CMakeLists.txt:454 (target_link_libraries):
  The keyword signature for target_link_libraries has already been used with
  the target "LIB_LIEF".  All uses of target_link_libraries with a target
  must be either all-keyword or all-plain.

  The uses of the keyword signature are here:

   * CMakeLists.txt:398 (target_link_libraries)
   * CMakeLists.txt:410 (target_link_libraries)
   * CMakeLists.txt:422 (target_link_libraries)
   * CMakeLists.txt:434 (target_link_libraries)

-- Configuration Types: 
-- Build Types: Debug
````

BehroozAbbassi commented:
> @BehroozAbbassi ,When I tested the features on my local machine with command "./vcpkg install lief[_] " and "./vcpkg install lief[_]:x64-windows" I met the error log:
> 
> ```
> CMake Error at CMakeLists.txt:454 (target_link_libraries):
>   The keyword signature for target_link_libraries has already been used with
>   the target "LIB_LIEF".  All uses of target_link_libraries with a target
>   must be either all-keyword or all-plain.
> 
>   The uses of the keyword signature are here:
> 
>    * CMakeLists.txt:398 (target_link_libraries)
>    * CMakeLists.txt:410 (target_link_libraries)
>    * CMakeLists.txt:422 (target_link_libraries)
>    * CMakeLists.txt:434 (target_link_libraries)
> 
> -- Configuration Types: 
> -- Build Types: Debug
> ```

I just test it and there is no error in my system 

First I removed all installed instances of the LIEF

```powershell
./vcpkg list | % { $_.Split(" ") | Select-Object -first 1 } | Where-Object { $_ -like "lief*:*" } | ForEach-Object -Process { ./vcpkg remove --recurse $_ }
```

And reinstall it with different features, everything is OK

```powershell
./vcpkg install lief[pe,elf,macho] --recurse

# Installation was successful

./vcpkg list | % { $_.Split(" ") | Select-Object -first 1 } | Where-Object { $_ -like "lief*:*" }                                                          ─╯
lief:x86-windows
lief[elf]:x86-windows
lief[macho]:x86-windows
lief[pe]:x86-windows
lief[shared-lib]:x86-windows
```

BehroozAbbassi committed:
Fix the runtime installation directory using a code patch

BehroozAbbassi committed:
"Update version database"

JonLiu1993 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg

JonLiu1993 commented:
@BehroozAbbassi , Thanks for your test, I have merged the master, could you please again?

JackBoosY commented on ports/lief/portfile.cmake:
```diff
@@ -0,0 +1,94 @@
+vcpkg_copy_pdbs()
+
+# # Handle copyright
+file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/LIEF" RENAME copyright)
```
```suggestion
```
I think we just need to install one copy here?

BehroozAbbassi replied:
> I think we just need to install one copy here?

I had to add the second one to fix an error with the `x64-linux` triplet.

JonLiu1993 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg

JonLiu1993 committed:
Merge branch 'master' of https://github.com/Microsoft/vcpkg

BehroozAbbassi committed:
Update ports/lief/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

BehroozAbbassi committed:
Update version database

JonLiu1993 commented:
@BehroozAbbassi , I tested again the feature on my local machine, but failed:

command:  ./vcpkg install lief[*]

````
PS F:\Feature-test\lief\vcpkg> ./vcpkg install lief[*]
Computing installation plan...
A suitable version of cmake was not found (required v3.22.2). Downloading portable cmake v3.22.2...
Downloading cmake...
  https://github.com/Kitware/CMake/releases/download/v3.22.2/cmake-3.22.2-windows-i386.zip -> F:\Feature-test\lief\vcpkg\downloads\cmake-3.22.2-windows-i386.zip
Extracting cmake...
The following packages will be built and installed:
    lief[art,asan,c-api,core,coverage,dex,disable-frozen,doc,elf,enable-json,examples,extra-warnings,force32,fuzzing,install-python,logging,logging-debug,lsan,macho,oat,pe,profiling,python-api,shared-lib,tests,tsan,usan,use-ccache,vdex]:x86-windows -> 0.11.5
  * vcpkg-cmake[core]:x64-windows -> 2022-01-19
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-windows...
A suitable version of powershell-core was not found (required v7.2.1). Downloading portable powershell-core v7.2.1...
Downloading powershell-core...
  https://github.com/PowerShell/PowerShell/releases/download/v7.2.1/PowerShell-7.2.1-win-x86.zip -> F:\Feature-test\lief\vcpkg\downloads\PowerShell-7.2.1-win-x86.zip
Extracting powershell-core...
Detecting compiler hash for triplet x86-windows...
A suitable version of 7zip was not found (required v21.7.0). Downloading portable 7zip v21.7.0...
Downloading 7zip...
  https://www.7-zip.org/a/7z2107-extra.7z -> F:\Feature-test\lief\vcpkg\downloads\7z2107-extra.7z
Extracting 7zip...
Restored 1 packages from C:\Users\vzhli17\AppData\Local\vcpkg\archives in 1.794 s. Use --debug to see more details.
Starting package 1/2: vcpkg-cmake:x64-windows
Installing package vcpkg-cmake[core]:x64-windows...
Elapsed time for package vcpkg-cmake:x64-windows: 83.55 ms
Starting package 2/2: lief:x86-windows
Building package lief[art,asan,c-api,core,coverage,dex,disable-frozen,doc,elf,enable-json,examples,extra-warnings,force32,fuzzing,install-python,logging,logging-debug,lsan,macho,oat,pe,profiling,python-api,shared-lib,tests,tsan,usan,use-ccache,vdex]:x86-windows...
-- Downloading https://github.com/lief-project/LIEF/archive/551ede538abeca63a158bd7c42b6b6337c92a26e.tar.gz -> lief-project-LIEF-551ede538abeca63a158bd7c42b6b6337c92a26e.tar.gz...
-- Extracting source F:/Feature-test/lief/vcpkg/downloads/lief-project-LIEF-551ede538abeca63a158bd7c42b6b6337c92a26e.tar.gz
-- Applying patch install_dir.patch
-- Using source at F:/Feature-test/lief/vcpkg/buildtrees/lief/src/337c92a26e-0a88e6d460.clean
-- Found external ninja('1.10.2').
-- Configuring x86-windows
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:128 (message):
    Command failed: ninja -v
    Working Directory: F:/Feature-test/lief/vcpkg/buildtrees/lief/x86-windows-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      F:\Feature-test\lief\vcpkg\buildtrees\lief\config-x86-windows-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:415 (vcpkg_execute_required_process)
  ports/lief/portfile.cmake:70 (vcpkg_cmake_configure)
  scripts/ports.cmake:145 (include)


Error: Building package lief:x86-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `git pull` and `.\vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+lief
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[lief]+Build+error
including:
  package: lief[art,asan,c-api,core,coverage,dex,disable-frozen,doc,elf,enable-json,examples,extra-warnings,force32,fuzzing,install-python,logging,logging-debug,lsan,macho,oat,pe,profiling,python-api,shared-lib,tests,tsan,usan,use-ccache,vdex]:x86-windows -> 0.11.5
    vcpkg-tool version: 2022-03-09-1affd32f93b299d5a907816c328ca3ededb73a7e
    vcpkg-scripts version: 2a0565e2c 2022-03-17 (12 hours ago)

Additionally, attach any relevant sections from the log files above.
````
F:\Feature-test\lief\vcpkg\buildtrees\lief\config-x86-windows-out.log

````
-- Undefined sanitizer enabled
CMake Error at CMakeLists.txt:454 (target_link_libraries):
  The keyword signature for target_link_libraries has already been used with
  the target "LIB_LIEF".  All uses of target_link_libraries with a target
  must be either all-keyword or all-plain.

  The uses of the keyword signature are here:

   * CMakeLists.txt:398 (target_link_libraries)
   * CMakeLists.txt:410 (target_link_libraries)
   * CMakeLists.txt:422 (target_link_libraries)
   * CMakeLists.txt:434 (target_link_libraries)
````

BehroozAbbassi commented:
> @BehroozAbbassi , I tested again the feature on my local machine, but failed:
> 
> command: ./vcpkg install lief[*]
> 
> ```
> PS F:\Feature-test\lief\vcpkg> ./vcpkg install lief[*]
> Computing installation plan...
> A suitable version of cmake was not found (required v3.22.2). Downloading portable cmake v3.22.2...
> Downloading cmake...
>   https://github.com/Kitware/CMake/releases/download/v3.22.2/cmake-3.22.2-windows-i386.zip -> F:\Feature-test\lief\vcpkg\downloads\cmake-3.22.2-windows-i386.zip
> Extracting cmake...
> The following packages will be built and installed:
>     lief[art,asan,c-api,core,coverage,dex,disable-frozen,doc,elf,enable-json,examples,extra-warnings,force32,fuzzing,install-python,logging,logging-debug,lsan,macho,oat,pe,profiling,python-api,shared-lib,tests,tsan,usan,use-ccache,vdex]:x86-windows -> 0.11.5
>   * vcpkg-cmake[core]:x64-windows -> 2022-01-19
> Additional packages (*) will be modified to complete this operation.
> Detecting compiler hash for triplet x64-windows...
> A suitable version of powershell-core was not found (required v7.2.1). Downloading portable powershell-core v7.2.1...
> Downloading powershell-core...
>   https://github.com/PowerShell/PowerShell/releases/download/v7.2.1/PowerShell-7.2.1-win-x86.zip -> F:\Feature-test\lief\vcpkg\downloads\PowerShell-7.2.1-win-x86.zip
> Extracting powershell-core...
> Detecting compiler hash for triplet x86-windows...
> A suitable version of 7zip was not found (required v21.7.0). Downloading portable 7zip v21.7.0...
> Downloading 7zip...
>   https://www.7-zip.org/a/7z2107-extra.7z -> F:\Feature-test\lief\vcpkg\downloads\7z2107-extra.7z
> Extracting 7zip...
> Restored 1 packages from C:\Users\vzhli17\AppData\Local\vcpkg\archives in 1.794 s. Use --debug to see more details.
> Starting package 1/2: vcpkg-cmake:x64-windows
> Installing package vcpkg-cmake[core]:x64-windows...
> Elapsed time for package vcpkg-cmake:x64-windows: 83.55 ms
> Starting package 2/2: lief:x86-windows
> Building package lief[art,asan,c-api,core,coverage,dex,disable-frozen,doc,elf,enable-json,examples,extra-warnings,force32,fuzzing,install-python,logging,logging-debug,lsan,macho,oat,pe,profiling,python-api,shared-lib,tests,tsan,usan,use-ccache,vdex]:x86-windows...
> -- Downloading https://github.com/lief-project/LIEF/archive/551ede538abeca63a158bd7c42b6b6337c92a26e.tar.gz -> lief-project-LIEF-551ede538abeca63a158bd7c42b6b6337c92a26e.tar.gz...
> -- Extracting source F:/Feature-test/lief/vcpkg/downloads/lief-project-LIEF-551ede538abeca63a158bd7c42b6b6337c92a26e.tar.gz
> -- Applying patch install_dir.patch
> -- Using source at F:/Feature-test/lief/vcpkg/buildtrees/lief/src/337c92a26e-0a88e6d460.clean
> -- Found external ninja('1.10.2').
> -- Configuring x86-windows
> CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:128 (message):
>     Command failed: ninja -v
>     Working Directory: F:/Feature-test/lief/vcpkg/buildtrees/lief/x86-windows-rel/vcpkg-parallel-configure
>     Error code: 1
>     See logs for more information:
>       F:\Feature-test\lief\vcpkg\buildtrees\lief\config-x86-windows-out.log
> 
> Call Stack (most recent call first):
>   installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:415 (vcpkg_execute_required_process)
>   ports/lief/portfile.cmake:70 (vcpkg_cmake_configure)
>   scripts/ports.cmake:145 (include)
> 
> 
> Error: Building package lief:x86-windows failed with: BUILD_FAILED
> Please ensure you're using the latest portfiles with `git pull` and `.\vcpkg update`.
> Then check for known issues at:
>   https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+lief
> You can submit a new issue at:
>   https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[lief]+Build+error
> including:
>   package: lief[art,asan,c-api,core,coverage,dex,disable-frozen,doc,elf,enable-json,examples,extra-warnings,force32,fuzzing,install-python,logging,logging-debug,lsan,macho,oat,pe,profiling,python-api,shared-lib,tests,tsan,usan,use-ccache,vdex]:x86-windows -> 0.11.5
>     vcpkg-tool version: 2022-03-09-1affd32f93b299d5a907816c328ca3ededb73a7e
>     vcpkg-scripts version: 2a0565e2c 2022-03-17 (12 hours ago)
> 
> Additionally, attach any relevant sections from the log files above.
> ```
> 
> F:\Feature-test\lief\vcpkg\buildtrees\lief\config-x86-windows-out.log
> 
> ```
> -- Undefined sanitizer enabled
> CMake Error at CMakeLists.txt:454 (target_link_libraries):
>   The keyword signature for target_link_libraries has already been used with
>   the target "LIB_LIEF".  All uses of target_link_libraries with a target
>   must be either all-keyword or all-plain.
> 
>   The uses of the keyword signature are here:
> 
>    * CMakeLists.txt:398 (target_link_libraries)
>    * CMakeLists.txt:410 (target_link_libraries)
>    * CMakeLists.txt:422 (target_link_libraries)
>    * CMakeLists.txt:434 (target_link_libraries)
> ```

hmm
I've the same problem by `lief[*]`, It's because of dependencies of some features, for instance the `doc` feature depends on `doxygen` and if cmake can't find the doxygen then build will fail.


But this one is related to sanitizers, I've no idea how to use them with MSVC and CMAKE. I'm not sure if it's an error with cmake script or trying to use sanitizers on windows, maybe the author of library could help here, @romainthomas  any idea?

```
-- Undefined sanitizer enabled
CMake Error at CMakeLists.txt:454 (target_link_libraries):
  The keyword signature for target_link_libraries has already been used with
  the target "LIB_LIEF".  All uses of target_link_libraries with a target
  must be either all-keyword or all-plain.

  The uses of the keyword signature are here:

   * CMakeLists.txt:398 (target_link_libraries)
   * CMakeLists.txt:410 (target_link_libraries)
   * CMakeLists.txt:422 (target_link_libraries)
   * CMakeLists.txt:434 (target_link_libraries)
 ```
 By the way I've this error even after I edit the `portfile.cmake` and disable sanitizers manually.
 
 ```cmake
 
 # Set them OFF after the "vcpkg_check_features"
 
set(LIEF_ASAN OFF)
set(LIEF_LSAN OFF)
set(LIEF_TSAN OFF)
set(LIEF_USAN OFF)
 
 ```

JonLiu1993 commented:
@romainthomas, Could you please take a look?

romainthomas commented:
I'll check but the error seems weird

romainthomas commented:
Maybe you should first try with the latest release (0.12.0) that contains CMake refactoring (cc @ekilmer)

ekilmer commented:
> the latest release (0.12.0) that contains CMake refactoring

I have a Draft PR for that here https://github.com/microsoft/vcpkg/pull/22957 (initially forked from @BehroozAbbassi's branch). The main issue is mbedtls major version upgrade in vcpkg (more details in the OP of the PR). I also haven't merged the latest changes (mostly need to just double-check that non-default features still work/should be removed) from this branch.

If the mbedtls version can't be updated for LIEF, then maybe LIEF will need to live in a separate [registry](https://devblogs.microsoft.com/cppblog/how-to-start-using-registries-with-vcpkg/) along with mbedtls v3.

I don't have time to work on anything this week, but anyone is free to fork my branch, make the necessary changes, and open a new PR, create a registry, whatever. Just let me know if https://github.com/microsoft/vcpkg/pull/22957 should be closed in favor of something else!

JackBoosY commented:
Just disable feature doc would be okay.

JonLiu1993 commented:
@BehroozAbbassi, Is there any progress this pr？

JonLiu1993 commented:
Draft this PR until it has any progress.
